### PR TITLE
Improve unity support

### PR DESCRIPTION
**Pull request by GwendalBroudin-Emoko**
* Add support for unity context object (use UnitySinkExtensions.WithUnityContext)
* Use UnityEngine.ILogger instead of Debug.log for better flexibility

_I know the dictionary implementation is not that great, but I didn't find another way to pass raw object to the sink for now_

### CODE REVIEW

**Comment by KuraiAndras** [comment]
So, a couple of things here:
1. I am not really seeing the usecase here. How is this feature supposed to be used? I (and I suppose most people) set up the logger once, around the start of your application and then don't create a new logger. With this are you supposed to set up a new instance of the logger for every `MonoBehaviour`?
2. Using switch expressions does not work with older Unity versions. Right now the package supports Unity version from 2019 and up, and right now I don't see the need to change that.
3. What does the Unity `ILogger` provide flexibility? If I understand it correctly, it should be a nice logging abstraction, and you can create you own implementation for it. Why would a Serilog library need this? If you want to log to multiple places other than the Unity debug log, then just use other Serilog sinks like the File sink, or create your own implementation.

Other than that, having the ability to pass a context object to the Unity logging mechanism, is something that would be nice.

File: Serilog.Sinks.Unity3D/Assets/Serilog.Sinks.Unity3D/Unity3DLogEventSink.cs [hunk 1]
```diff
@@ -1,44 +1,75 @@
...
+                    LogEventLevel.Verbose
+                    or LogEventLevel.Debug
+                    or LogEventLevel.Information => LogType.Log,
+
+                    LogEventLevel.Warning => LogType.Warning,
+
+                    LogEventLevel.Error or LogEventLevel.Fatal => LogType.Error,
+
+                    _ => throw new ArgumentOutOfRangeException(nameof(logEvent.Level), "Unknown log level")
+                };
```
Comment (KuraiAndras) [inline Serilog.Sinks.Unity3D/Assets/Serilog.Sinks.Unity3D/Unity3DLogEventSink.cs]: Not compatible with Unity 2019

**Review by KuraiAndras — CHANGES_REQUESTED** [review]

**Comment by GwendalBroudin-Emoko** [comment]
The use of ILogger instead of Debug.Log... is just for custom implementation of [Unity Logger](https://docs.unity3d.com/ScriptReference/ILogger.html) and Debug.Log actually use Unity's default implementation Debug.unityLogger.

I recall that some custom logging UI for Unity use this with [ILogHandler](https://docs.unity3d.com/ScriptReference/ILogHandler.html), this change facilitates the integration of those packages.

**Comment by GwendalBroudin-Emoko** [comment]
Hi @KuraiAndras, could you review the change I made?

**Comment by KuraiAndras** [comment]
Sorry, I will try to take a look later today

File: Serilog.Sinks.Unity3D/Assets/Serilog.Sinks.Unity3D/UnitySinkExtensions.cs [hunk 2]
```diff
@@ -52,12 +53,23 @@ public static class UnitySinkExtensions
...
-            return sinkConfiguration.Sink(new Unity3DLogEventSink(formatter), restrictedToMinimumLevel, levelSwitch);
+            return sinkConfiguration.Sink(new Unity3DLogEventSink(formatter, logger), restrictedToMinimumLevel, levelSwitch);
+        }
+
+
+        public static ILogger WithUnityContext(this ILogger logger, UnityEngine.Object context)
+        {
+            int id = context.GetInstanceID();
+#if UNITY_EDITOR
+            Unity3DLogEventSink._objectstoLog.TryAdd(id, context);
```
Comment (KuraiAndras) [inline Serilog.Sinks.Unity3D/Assets/Serilog.Sinks.Unity3D/UnitySinkExtensions.cs:70]: With Unity 2019.2.0 the Dictionary<T,R> type does not have the TryAdd method, and this does not compile in the editor
image

**Comment by KuraiAndras** [comment]
So, how would you intend to use the `WithUnityContext` method? Every user script creates their own instance of the `ILogger`? Also, not really a fan of using a static dictionary to store `UnityObject`s

File: Serilog.Sinks.Unity3D/Assets/Serilog.Sinks.Unity3D/UnitySinkExtensions.cs [hunk 3]
```diff
@@ -52,12 +53,23 @@ public static class UnitySinkExtensions
             this LoggerSinkConfiguration sinkConfiguration,
             ITextFormatter formatter,
             LogEventLevel restrictedToMinimumLevel = LevelAlias.Minimum,
-            LoggingLevelSwitch levelSwitch = null)
+            LoggingLevelSwitch levelSwitch = null,
+            UnityEngine.ILogger logger = null)
```
Comment (KuraiAndras) [inline Serilog.Sinks.Unity3D/Assets/Serilog.Sinks.Unity3D/UnitySinkExtensions.cs:57]: If we add an ILogger parameter to this method, then why not support ILogger on the other method too?

File: Serilog.Sinks.Unity3D/Assets/Serilog.Sinks.Unity3D/UnitySinkExtensions.cs [hunk 4]
```diff
@@ -52,12 +53,23 @@ public static class UnitySinkExtensions
...
         {
             if (sinkConfiguration == null) throw new ArgumentNullException(nameof(sinkConfiguration));
             if (formatter == null) throw new ArgumentNullException(nameof(formatter));
 
-            return sinkConfiguration.Sink(new Unity3DLogEventSink(formatter), restrictedToMinimumLevel, levelSwitch);
+            return sinkConfiguration.Sink(new Unity3DLogEventSink(formatter, logger), restrictedToMinimumLevel, levelSwitch);
+        }
+
+
+        public static ILogger WithUnityContext(this ILogger logger, UnityEngine.Object context)
```
Comment (KuraiAndras) [inline Serilog.Sinks.Unity3D/Assets/Serilog.Sinks.Unity3D/UnitySinkExtensions.cs:66]: This method should have comments in a similar fashing as on the other overloads

**Review by KuraiAndras — CHANGES_REQUESTED** [review]
I am not comfortable with the usage of the static dictionary for logging the context objects, also the whole thing does not compile (even after fixing the switch expressions), so you probably did not open this with the intended Unity version.

Also, Please provide sample usage of how would providing an object context look like with multiple scripts

**Comment by KuraiAndras** [comment]
Even though the PR cannot be compiled in Unity 2019, I have started working on updating the package to Unity 2021.3, as the newer versions of the visual studio integration do not support the 2019 version.

**Comment by KuraiAndras** [comment]
Even though the PR cannot be compiled in Unity 2019, I have started working on updating the package to Unity 2021.3, as the newer versions of the visual studio integration do not support the 2019 version.

**Comment by KuraiAndras** [comment]
This feature (with others) is included in the 3.0.0 version. Thank you for bringing this to my attention and for providing the initial implementation!

**Pull request closed without merging**